### PR TITLE
change: use `python:X.Y-slim` base Docker image in place of `python:X.Y` images

### LINKED ISSUE #3203: Slim Meltano Docker images
Migrated from GitLab: https://gitlab.com/meltano/meltano/-/issues/3278

Originally created by @Stephen.Clarke.health.govt.nz on 2022-02-23 07:41:11

---

### Problem to solve
Reducing security vulnerabilities in Meltano Docker Images to meet IT Security Guidelines.

### Target audience
This will be beneficial for all Meltano Users who use the Meltano Docker Images - especially helpful for Containerized implementations of Meltano.

### Further details
When scanning the Meltano Image for vulnerabilities (via AWS Container Registry - ECR), there are many security vulnerabilities listed. They mainly appear to be attributed to the Docker Python base image.

My security team would like to minimize the number of identified vulnerabilities, and the only way to do this is to change the python base image Meltano is built on.  

### Proposal
Having come across this article, https://snyk.io/blog/best-practices-containerizing-python-docker/, they appear to recommend alternative base images for Python. For example I am building a Meltano Python 3.8 Image, the recommendation is
python:3.8.12-slim-bullseye
I'm wonder if it is possible to have alternative Meltano Docker builds which I could pull with a more hardened base image?

I have tested and use the slim-bullseye by making my own Custom Meltano Base image and it seems to work well. The other advantage is that it reduces the size of the overall Docker Image.

Please Note: I really don't want to maintain my own custom meltano base image as I need to update a number of the other dependency files e.g. pyproject.toml. This makes my builds brittle and subject to build failure when there are updated Meltano releases.

### What does success look like, and how can we measure that?
Success for me would be less security vunerabilities.

I feel this could be done either one of two ways which aligns to @aaronsteers thinking as well.

1. Completely swap out the Python Base image with the slim Python versions as suggested by the Synk article.
2. Have alternative builds / variants Meltano Docker Images which use the slim Python image for the Meltano Base build.

### Regression test
(Ensure the feature doesn't cause any regressions)
- [ ] Regression build some Meltano Docker images to ensure TAPS and TARGETS still work
- [ ] Consider a phased deployed e.g. Option 2 first and then Option 1. In that way users can confirm the new slim python builds work for them.

### Links / references

https://snyk.io/blog/best-practices-containerizing-python-docker/

The attached file is my example [Dockerfile](http://gitlab.com/meltano/meltano/uploads/435de894cd04eef11959b31a7b3c7a4d/Dockerfile) for my custom based build of the Meltano Base using a slim-python base image.

### PR DESCRIPTION
**Pull request by WillDaSilva**
The main change in this PR is that we now base our Docker images off of `python:<version>-slim` instead of `python:<version>`.

This reduces the number of vulnerabilities Grype detects with the image from 1173 to 90, most of which have a severity of "negligible". The high/critical severity issue are listed below:

```
e2fsprogs         1.46.2-2            (won't fix)  deb   CVE-2022-1304     High
libcom-err2       1.46.2-2            (won't fix)  deb   CVE-2022-1304     High
libdb5.3          5.3.28+dfsg1-0.8    (won't fix)  deb   CVE-2019-8457     Critical
libext2fs2        1.46.2-2            (won't fix)  deb   CVE-2022-1304     High
libgcrypt20       1.8.7-6             (won't fix)  deb   CVE-2021-33560    High
libncursesw6      6.2+20201114-2      (won't fix)  deb   CVE-2022-29458    High
libpcre2-8-0      10.36-2             (won't fix)  deb   CVE-2022-1586     Critical
libpcre2-8-0      10.36-2             (won't fix)  deb   CVE-2022-1587     Critical
libss2            1.46.2-2            (won't fix)  deb   CVE-2022-1304     High
libssl1.1         1.1.1n-0+deb11u3    (won't fix)  deb   CVE-2022-2097     High
libtinfo6         6.2+20201114-2      (won't fix)  deb   CVE-2022-29458    High
logsave           1.46.2-2            (won't fix)  deb   CVE-2022-1304     High
ncurses-base      6.2+20201114-2      (won't fix)  deb   CVE-2022-29458    High
ncurses-bin       6.2+20201114-2      (won't fix)  deb   CVE-2022-29458    High
openssl           1.1.1n-0+deb11u3    (won't fix)  deb   CVE-2022-2097     High
perl-base         5.32.1-4+deb11u2    (won't fix)  deb   CVE-2020-16156    High
```

They are all marked as "won't fix", unfortunately. I went through each one and tried to find a way to update or remove it safely, but there were no easy wins. Even so, this change provides a significant benefit in terms of both image size (down from 1.1 GB to 293 MB) and security.

Relates to:
- #6498

Closes #3203

### CODE REVIEW

**Review by pandemicsyn — APPROVED** [review]
lgmt. @aaronsteers do we need to give users a heads-up on this?

**Comment by aaronsteers** [comment]
@WillDaSilva - Should we be targeting `main` with this PR? Can we switch this PR target or if not, can you say more about what we need from the other branch to be deployed at the same time?

**Comment by WillDaSilva** [comment]
> @WillDaSilva - Should we be targeting `main` with this PR? Can we switch this PR target or if not, can you say more about what we need from the other branch to be deployed at the same time?

@aaronsteers I've updated this PR to target `main`.

**Comment by WillDaSilva** [comment]
- https://github.com/meltano/meltano/pull/6555

^ This will need to be merged in before CI passes for this PR.

File: docker/meltano/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,27 @@
+ARG PYTHON_VERSION
+
+FROM python:${PYTHON_VERSION}-slim as build
```
Comment (aaronsteers) [inline docker/meltano/Dockerfile:3]: Nit: what do you think of dropping the `as build` here? Whenever I see a name here I expect a multi-build process, which we don't have here.

```suggestion
FROM python:${PYTHON_VERSION}-slim
```

If there's an advantage of having this clause added, I don't mind keeping it.

**Comment by aaronsteers** [comment]
> lgmt. @aaronsteers do we need to give users a heads-up on this?

Yeah, good call. I've updated the PR to a semantic description which should give the context in the release notes.

**Review by aaronsteers — APPROVED** [review]
Looks great! 🚢 

One tiny nit (non-blocking).

File: docker/meltano/Dockerfile (hunk 1 above)
Comment (WillDaSilva) [reply docker/meltano/Dockerfile:3]: Good catch. That's a holdover from when I was doing this with a multi-stage build. I'll remove it.

[the remark this answers is not in the document]
Comment (WillDaSilva) [reply docker/meltano/Dockerfile]: We could, but as the Dockerfile gets more complicated to patch security issues (see #6497) it'll become more likely that the recipe will only work for the `-slim` base images, and so the `BASE_IMAGE_SUFFIX` might become useless or misleading.

If we want to maintain compatibility with both slim and non-slim base images, that can be done, but we should be explicit about it, and note it in the Dockerfile (and possibly in the handbook too).

[the remark this answers is not in the document]
Comment (aaronsteers) [reply docker/meltano/Dockerfile]: Yeah. Makes sense! 👍

**Comment by WillDaSilva** [comment]
> * [ci: Pin `fossa-cli` version to `$FOSSA_CLI_VERSION` & update to v3.3.11 #6555](https://github.com/meltano/meltano/pull/6555)
> 
> 
> ^ This will need to be merged in before CI passes for this PR.

Looks like `fossa-cli` must've done something to fix the problem without us having to change anything. Or maybe the problem was with GitHub all along? In any case, this PR is no longer blocked!

**Comment by s7clarke11** [comment]
Nice work @WillDaSilva , great outcome (thank for putting a PR together for this). This change will make my security team so much happier. I'm pleased I don't need to maintain my own custom Meltano base image now.

**Pull request merged**